middleware/signmigration.py

```diff
@@ -0,0 +1,246 @@
+    if options.output_path is None:
+        info(eth_message_to_printable(migration_spec.get_authorization_msg()))
+    else:
+        sgx_authorization.save_to_jsonfile(options.output_path)
+        info(f"SGX migration authorization saved to {options.output_path}")
+
+
+def do_manual_sign(options):
+    _require_output_path(options, require_existing=True)
+    if options.signature is None:
+        raise AdminError("Must provide a signature (-g/--signature)")
+
+    info(f"Opening SGX migration authorization file {options.output_path}...")
+    sgx_authorization = SGXMigrationAuthorization.from_jsonfile(options.output_path)
+    info("Adding signature...")
+    sgx_authorization.add_signature(options.signature)
+    sgx_authorization.save_to_jsonfile(options.output_path)
+    info(f"SGX migration authorization saved to {options.output_path}")
+
+
+def do_key(options):
+    _require_output_path(options, require_existing=True)
+    if options.key is None:
+        raise AdminError("Must provide a signing key (-k/--key)")
+    if not is_hex_string_of_length(options.key, 32, allow_prefix=True):
+        raise AdminError(f"Invalid key '{options.key}'")
+
+    info(f"Opening SGX migration authorization file {options.output_path}...")
+    sgx_authorization = SGXMigrationAuthorization.from_jsonfile(options.output_path)
+    migration_spec = sgx_authorization.migration_spec
+    info("Signing with key...")
+    sk = ecdsa.SigningKey.from_string(
+        bytes.fromhex(options.key),
+        curve=ecdsa.SECP256k1
+    )
+    signature = sk.sign_digest(
+        migration_spec.get_authorization_digest(),
+        sigencode=ecdsa.util.sigencode_der
+    )
+    # Add the signature to the authorization and save it to disk
+    sgx_authorization.add_signature(signature.hex())
+    sgx_authorization.save_to_jsonfile(options.output_path)
+    info(f"SGX migration authorization saved to {options.output_path}")
+
+
+def do_eth(options):
+    _require_output_path(options)
+    if options.path is None:
+        options.path = DEFAULT_ETH_PATH
+    # Parse path
+    path = BIP32Path(options.path)
+    eth = None
+    try:
+        # Get dongle access (must have ethereum app open)
+        eth = get_eth_dongle(options.verbose)
+        # Retrieve public key
+        info(f"Retrieving public key for path '{str(path)}'...")
+        pubkey = eth.get_pubkey(path)
+        info(f"Public key: {pubkey.hex()}")
+
+        # If options.pubkey is True, we just want to retrieve the public key
+        if options.pubkey:
+            info(f"Opening public key file {options.output_path}...")
+            info("Adding public key...")
+            with open(options.output_path, "w") as file:
+                file.write("%s\n" % pubkey.hex())
+            info(f"Public key saved to {options.output_path}")
+            return
+
+        # Is there an existing migration authorization? Read it
+        sgx_authorization = None
+        if not isfile(options.output_path):
```

Comment:
Don't we have a function to check this?

Reply:
Nice catch, yes, I'll replace it with the function.